### ISSUE TITLE
Investigate Django for RFSwarm's V2 UI

### ISSUE THREAD

**Issue by damies13** [damies13/rfswarm#317]
In Issue #247, @Krata4  has suggested using the Django Framework.

This issue is to investigate more about this suggestion, to understand it better and see if it will be a good fit for RFSwarm

**Comment by damies13** [comment]
@Krata4 To continue on from our discussion in Issue #247, I watched a couple of YT videos on Django to get a better grasp of what it is:
- https://youtu.be/cJveiktaOSQ?si=BmjK_W7myAQqWeOc
- https://youtu.be/NoLF7Dlu5mc?si=opJp81lrY_v7kFKG

Let me know if there are better video's than this?

  - As I need to re-write the API's anyway to work with eel I'm open to using the Django framework, and before we start a re-write of the API's is the best time to make these kind of design decisions, so thank you for the suggestion. 
  - I saw that Django uses serialises and this did all the validation of the post/put requests, I see that's really useful and likely better than what I have currently

After watching these videos I have many questions, I'll start with the most important ones:

1. I can see I'm going to need the help of someone already familiar with Django in order to integrate it with RFSwarm, are you willing and able to be the Django "expert" for this project?
2. Both video's showed creating a project  and then running the main project file to start a Django server, but you mentioned that Django can be run in Flask? is that the same thing as the bottles that eel is using?
3. In the 1st video it didn't show where the data was written, the 2nd video showed a sqlite3 file but didn't show the data that was written to the file and neither showed the SQL, so I get the impression that Django is providing an abstraction layer from the data, I'm just not sure how that will work with the data structure that RFSwarm already has? I don't want the data structure changed, so does Django require it's own data structures or is that just a limitation of the simplicity of these examples in these videos?

**Comment by Krata4** [comment]
@damies13 I think Django has a lot of documentations and videos :-)

ad1: I am developing at Django Framework 2 years. I do not know if I am expert :-), but I am ready to help.
ad2: Django supporting apps which you can install. So, we can create app for Django framework as wraper.  For Flask you can include Library only and use it.
ad3: Django is supporting any data model. You have to develop layer with you data model and provide as interface to Django.

example of plug-in (app) for Django:
https://docs.django-cms.org/en/release-3.9.x/how_to/custom_plugins.html
there is descibe how to define yourown data model from plug-in app

**Comment by Krata4** [comment]
@damies13  good video: 
https://www.youtube.com/watch?v=rHux0gMZ3Eg

You saw video for REST API only. But we do not want create REST API, I hope :-)

**Comment by damies13** [comment]
BTW are you Jiri Kratochvil on Slack?

**Comment by Krata4** [comment]
yes :-)

**Comment by damies13** [comment]
Some more questions:
- I notice that all the video's always start with `django-admin startproject <name>` do we need to do this? I'm wondering how this will fit in with the existing application?
- I was a bit surprised by your comment "But we do not want create REST API, I hope", The existing api was based on the res model, though admittedly it's only a partial rest api, it only has 1 get call, and the rest are all post calls, it's probably not a true rest api. I do need to reproduce this API as the agent relies on it for communication with the manager, so this would be the first thing to write in django. If your curious [this page](https://github.com/damies13/rfswarm/blob/master/Doc/Agent_Communication.md) documents the api we need to make available to the agents with example post and response bodies.
- I'm still not clear how django hooks into an existing application, so I'll start a 2.0.0 branch and set it up with the beginnings of the eel components that I do understand, then I'll ask for your help to hook django in to it

**Comment by damies13** [comment]
There are some other big prerequisite tasks that I think need to happen for the v2 RFSwarm .... the most important one, currently the 3 parts, Manager, Agent and Reporter are all 3 monolithic python files with several classes in those files, there is also some significant code duplication between them, I've known for quite a while this needs to be addressed, but was worried about breaking things when breaking out the classes. 

This is why the last year has been a major focus on getting the regression suite built, once the regression suite has good coverage of the applications I can work on breaking the classes.out to their own files and sharing them between the apps where possible, I see this is as a major prerequisite step on the path to the v2 release, and I've been hesitant to start the v2 code base before this was done

I think it's important to undo the mistakes of the past before moving forward to the nice to have features.

**Comment by damies13** [comment]
I kind of mentioned it before, we actually need to re-write the current api. (sorry this is a bit long with some history)

I'm not really familiar with all the high level stuff, this is probably because I came into IT from an electronics background, my first programming was assembler for PLC's and EEPROMS, using assembler to program them to emulate chips that could replace logic circuits made from transistors. later I learned to write C code my first "high level" language.

When I started RFSwarm, I didn't know much about the frameworks available for python, so I did what I knew best  and started at the protocol level, saw that python has a basic http server builtin so used that, well the [ThreadingHTTPServer)](https://docs.python.org/3/library/http.server.html#http.server.ThreadingHTTPServer).

Actually everything I did when starting RFSwarm was done preferring the modules that get installed with the base python install rather than other modules, e.g. the gui is written with pure tkinter, not some nice user friendly module like wxwidgets that uses tkinter, just the base tkinter module (it's a common theme)

When I wanted to move to a html based UI I still wanted the desktop style app, I don't want it to be a pure client server app in the sense of the traditional web app, but rather something more like an electron style app but one where you could also point a browser at it from a remote machine as well, that's where eel and issue #247 came in.

The problem with eel is it needs to run it's own web server (or maybe it can be run under an existing a bottles environment?) and can't use the basic http.server that comes builtin in python that I was using. I don't want to have the manager open several web server ports when the app runs, I want it to remain simple for the end user, 1 single port means 1 firewall rule, no confusion and trying to connect the agent to the wrong port etc.

So because of this, we need to rewrite the API, so we may as well start here with django, and then add other api's as needed keeping all api's in django so everything is consistent. at least that's where my thinking is currently.

**Comment by damies13** [comment]
Since I closed this issue Eel has stopped being supported and is no longer maintained, so the primary reason for not using Django no longer exists.

My current thoughts are to re-open this issue and in an early build of RFSwarm V2 build a partial V1 api in Django, Bottles, Flask, and any other Python API Libraries and do a basic comparison:

- Compare how they perform vs the baseline builtin ThreadingHTTPServer
- Compare how easy it is to write api end points with each library
- Compare how heavy the overhead of each library is (e.g. how many python packages they depend on)

After comparing the various libraries then make a decision on which one to go with.

**Comment by damies13** [comment]
Another to add to the list for consideration:
- FastAPI
or 
- FastAPI + uvicorn

Example [here](https://github.com/roshanlam/sqlite-worker/tree/main/examples/fastapi_integration)

Now we have a baseline test we can build variants to test against

**Comment by damies13** [comment]
Status Update:

- I've created test cases that let us measure the response times of the API calls
- I've added multiprocessing queue handling functions with the core logic behind the API calls, so now we can implement various webserver options so we can compare them fairly.


## Baseline - 1 September 2026
- Run [#1581](https://github.com/damies13/rfswarm/actions/runs/33448017936)
- This was run on the initial v2.0.0 build after merging v1.6.0 into the v2.0.0 branch but before making any changes related to v2.0.0

### Ubuntu with Python 3.13

Let's start with a table showing the full test results:

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 0.007 | 0.009 | 0.012  | 1.015 | 0.012 | 7764 | 0    | 0     |
|  Check ${resp_result_1} Contains Agents Name    | 0.002 | 0.002 | 0.003  | 0.024 | 0.001 | 5803 | 0    | 0     |
|  Simulated Keyword Name                         | 0.01  | 0.036 | 0.081  | 0.091 | 0.027 | 5803 | 0    | 0     |
|  Check ${resp_result_1} Has Result Queued       | 0.003 | 0.003 | 0.005  | 0.011 | 0.001 | 5803 | 0    | 0     |
|  Send Result to Manager                         | 0.02  | 0.026 | 0.031  | 1.028 | 0.014 | 5803 | 0    | 0     |
|  Check ${resp_result} Has Metric my_test_server | 0.003 | 0.004 | 0.005  | 0.019 | 0.001 | 1961 | 0    | 0     |
|  Check ${resp_result} Has Result Queued         | 0.003 | 0.003 | 0.005  | 0.008 | 0.001 | 1961 | 0    | 0     |
|  Send Metric to Manager                         | 0.021 | 0.026 | 0.032  | 0.051 | 0.004 | 1961 | 0    | 0     |

The result we'll be using for comparison as our baseline is:

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 0.007 | 0.009 | 0.012  | 1.015 | 0.012 | 7764 | 0    | 0     |

We'll be comparing the 90%ile result and the pass count as the test will always run for the same time, the lower 90%ile is better and the higher Pass count is better

### Ubuntu with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 0.007 | 0.009 | 0.01   | 0.028 | 0.001 | 7787 | 0    | 0     |

### MacOS with Python 3.13

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 0.004 | 0.017 | 0.02   | 2.059 | 0.078 | 7090 | 0    | 0     |

### MacOS with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 0.004 | 0.02  | 0.02   | 3.027 | 0.107 | 7050 | 0    | 0     |

### Windows with Python 3.13

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 2.012 | 2.028 | 2.04   | 2.056 | 0.009 | 912  | 0    | 0     |

### Windows with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| -------------------------------------------------|---------|---------|----------|----------|----------|--------|----|-----|
|  Send POST Request To the Manager               | 2.013 | 2.029 | 2.041  | 2.067 | 0.01  | 911  | 0    | 0     |

**Comment by damies13** [comment]
## Baseline - 16 September 2026

- Run [#1605](https://github.com/damies13/rfswarm/actions/runs/34979626356)
 - This run was after adding multiprocessing queue handling functions with the core logic behind the API calls, the multiprocessing queue handling functions are not used, still using the original BaseHTTPRequestHandler API

### Ubuntu with Python 3.13

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 0.006 | 0.01  | 0.013  | 0.042 | 0.002 | 7688 | 0    | 0     |

### Ubuntu with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 0.006 | 0.008 | 0.01   | 0.021 | 0.002 | 7748 | 0    | 0     |

### MacOS with Python 3.13

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 0.004 | 0.014 | 0.015  | 2.012 | 0.072 | 7201 | 0    | 0     |

### MacOS with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 0.003 | 0.019 | 0.021  | 3.031 | 0.099 | 6946 | 0    | 0     |

### Windows with Python 3.13

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 2.008 | 2.026 | 2.038  | 2.062 | 0.009 | 916  | 0    | 0     |

### Windows with Python 3.14

|  Result Name                                    | Min   | Avg   | 90%ile | Max   | Stdev | Pass | Fail | Other |
| ------------------------------------------------ | ------- | ------- | ---------- | -------- | -------- | ------- | ------ | ------------- | 
|  Send POST Request To the Manager               | 2.01  | 2.027 | 2.038  | 2.07  | 0.009 | 914  | 0    | 0     |
